### PR TITLE
Enable SSR for Plugins landing page

### PR DESCRIPTION
**Pull request by claudiucelfilip**
#### Proposed Changes

More info: p58i-d84-p2#comment-55362
This enables SSR for the Plugins landing page route (/plugins). It contains a part from #66344 (reverted in #67573) and hides changes behind a `plugins/ssr-landing` flag.

#### Testing Instructions

* Log out and disable JavaScript
* Got to `/plugins`
* You should see the Plugins landing page with the 3 sections of Plugins: Paid, Editor's Pick and Free (Note, breadcrumbs are not shown)
Screenshot on 2022-09-13 at 18-00-51

* To disable the feature flag run: `DISABLE_FEATURES=plugins/ssr-landing yarn start `
* Refreshing the `/plugins` page should  only show the WordPress placeholder logo

Screenshot on 2022-09-13 at 17-45-33 (1)

#### Pre-merge Checklist

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

Related to #

### CODE REVIEW

**Comment by jsnajdr** [comment]
I'm starting to look at this recent SSR work after returning from sabbatical, and initially I just wanted to ask: is this PR mostly re-adding the stuff removed in the revert PR (#67573), plus a feature flag?

File: client/data/marketplace/use-wpcom-plugins-query.ts [hunk 1]
```diff
@@ -37,6 +37,17 @@ const fetchWPCOMPlugins = ( type: Type, searchTerm?: string, tag?: string ) => {
 	);
 };
 
+export const getFetchWPCOMPlugins = (
+	enabled: boolean,
+	type: Type,
+	searchTerm?: string,
+	tag?: string
+): [ QueryKey, QueryFunction< { results: Array< { [ plugin: string ]: Plugin } > } > ] => {
+	const cacheKey = getCacheKey( type + searchTerm + tag + `${ enabled ? 'enabled' : 'disabled' }` );
```
Comment (jsnajdr) [inline client/data/marketplace/use-wpcom-plugins-query.ts]: Why would the `enabled` flag be a part of the query key? The `enabled` flag is used to make the `useQuery` hook into a "noop" conditionally, as the hooks itself must always be called.

File: client/my-sites/plugins/controller-logged-out.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+function prefetchPluginsData( queryClient, fetchParams, infinite ) {
+	const queryType = infinite ? 'fetchInfiniteQuery' : 'fetchQuery';
+
+	return queryClient[ queryType ]( ...fetchParams, {
+		enabled: true,
+		staleTime: BASE_STALE_TIME,
+		refetchOnMount: false,
+	} ).catch( () => {} );
```
Comment (jsnajdr) [inline client/my-sites/plugins/controller-logged-out.js]: `QueryClient` has special methods for prefetching: `queryClient.prefetchQuery` and `queryClient.prefetchInfiniteQuery`. The params are exactly the same, but the return value is `Promise< void >`: doesn't return the data, and also never fails, so the `.catch` handler can be removed.

The `enable: true` flag is redundant, as the query is enabled by default. Also, the `refetchOnMount` is not needed? Here we are not in React at all, we're calling the OOP API of `QueryClient` outside React. The `refetch*` params don't make sense here: the fetch is done only once, there's no React component that would live for an extended period of time and do refetches when needed.

File: client/data/marketplace/use-wpcom-plugins-query.ts (hunk 1 above)
Comment (claudiucelfilip) [reply client/data/marketplace/use-wpcom-plugins-query.ts]: That makes sense. I copied over this way from where it was in `useWPCOMPlugins` and avoided tinkering with the details.

**Comment by claudiucelfilip** [comment]
@jsnajdr  Thanks for looking at this. 
> is this PR mostly re-adding the stuff removed in the revert PR (#67573), plus a feature flag?

Indeed, this is the first step and it contains only a part of the initial SSR PR as it ended up too big to be properly evaluated.
Therefore, I added the feature flag until things ready for production.

Follow-up PR with be opened for the rest of the changes plus a few improvements to avoid future crashes (outlined p58i-d84-p2#comment-55361)

File: client/my-sites/plugins/controller-logged-out.js (hunk 2 above)
Comment (claudiucelfilip) [reply client/my-sites/plugins/controller-logged-out.js]: Thanks for the suggestions. I removed the unnecessary bits.
I can't remember why I thought one of the  `prefetch*` calls didn't work. I switched now and they seem to work fine.

**Review by worldomonation — APPROVED** [review]
Approving on behalf of KitKat for changes under `test/e2e` and `packages/calypso-e2e`.

File: client/my-sites/plugins/controller-logged-out.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+		tag: '',
+	};
+	return props;
+}
+
+function prefetchPluginsData( queryClient, fetchParams, infinite ) {
+	const queryType = infinite ? 'prefetchInfiniteQuery' : 'prefetchQuery';
+
+	return queryClient[ queryType ]( ...fetchParams, {
+		staleTime: BASE_STALE_TIME,
```
Comment (jsnajdr) [inline client/my-sites/plugins/controller-logged-out.js]: This `staleTime` property, I think it could be eliminated from the prefetch call and also from all the `useQuery` hook calls. It could be set once and for all when the `QueryClient` is being created:
```js
queryClient = new QueryClient( {
  defaultOptions: {
    queries: {
      staleTime: BASE_STALE_TIME,
    },
  },
} );
```
Because why would we need to customize it for any individual query?

The `QueryClient` configurations for client and server should be different:
- client: `staleTime` is 0, the default, meaning that the data are stale immediately after fetching. A `useQuery` hook will return the stale data to be displayed in UI, and then refetch immediately to get the fresh data (according to flags like `refetchOnMount`.
- server: `staleTime` is some long time like 2 hours, meaning that the `prefetchQuery` call will fetch only when the data are more than 2 hours old, otherwise it's noop. That how caching is achieved.

File: client/my-sites/plugins/controller-logged-out.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+	} );
+}
+
+const prefetchPaidPlugins = ( queryClient, options ) =>
+	prefetchPluginsData(
+		queryClient,
+		getFetchWPCOMPlugins( true, 'all', options.search, options.tag )
```
Comment (jsnajdr) [inline client/my-sites/plugins/controller-logged-out.js]: The `getFetchWPCOMPlugins` function returns _parameters_ for the query (key and fn), but that's not communicated by the function name at all. I would call it something like `getWPCOMPluginsQueryParams` instead.

Also, the first parameter (`enabled`) can be removed now when `enabled` is no longer part of the query key.

File: client/data/marketplace/use-wporg-plugin-query.ts [hunk 5]
```diff
@@ -18,35 +20,41 @@ import { BASE_STALE_TIME, WPORG_CACHE_KEY } from './constants';
...
+			author,
+			tag: options.tag && ! search ? options.tag : null,
+		} );
+	};
+	return [ cacheKey, fetchFn ];
+};
+
 export const useWPORGPlugins = (
 	options: PluginQueryOptions,
 	{ enabled = true, staleTime = BASE_STALE_TIME, refetchOnMount = true }: UseQueryOptions = {}
```
Comment (jsnajdr) [inline client/data/marketplace/use-wporg-plugin-query.ts:45]: I believe there is a better way to pass custom query options here. Instead of supporting just two (somewhat randomly chosen) options, why not allow to customize everything? A common pattern that we've been using in other Calypso queries (e.g., [`useFollowersQuery`](https://github.com/Automattic/wp-calypso/blob/trunk/client/data/followers/use-followers-query.js)) is like:
```js
function useMyQuery( myQueryOptions, queryOptions) {
  return useQuery( key, fn, {
    ...queryOptions,
    select: ( data ) => { ... },
  } );
}
```
I.e., just spread the entire `queryOptions` object and add the non-overridable options like `select`. That should be more concise and more universal.

File: packages/search/src/search.tsx [hunk 6]
```diff
@@ -419,11 +419,16 @@ const InnerSearch = (
 			return renderReskinSearchIcon();
 		}
 
+		let focusHandler;
+		if ( 'undefined' !== typeof window ) {
+			focusHandler = window.focus;
+		}
```
Comment (jsnajdr) [inline packages/search/src/search.tsx]: This is a great catch but the fix is not correct. We didn't intend to call `window.focus` here, it's a reference to what used to be a `this.focus` method in a class component and was forgotten during refactoring to functional component.

I filed a better fix in #67800. Let's merge that and then rebase this PR on top, without modifying the `Search` component further.

File: client/my-sites/plugins/controller-logged-out.js (hunk 3 above)
Comment (claudiucelfilip) [reply client/my-sites/plugins/controller-logged-out.js]: Fair point. 
`staleTime` removed.

File: client/my-sites/plugins/controller-logged-out.js (hunk 4 above)
Comment (claudiucelfilip) [reply client/my-sites/plugins/controller-logged-out.js]: Agreed and renamed them the functions.

File: client/data/marketplace/use-wporg-plugin-query.ts (hunk 5 above)
Comment (claudiucelfilip) [reply client/data/marketplace/use-wporg-plugin-query.ts:45]: Sounds good. 
I'll address this tomorrow as this seems to trigger some type mismatches for the `select` function.

File: packages/search/src/search.tsx (hunk 6 above)
Comment (claudiucelfilip) [reply packages/search/src/search.tsx]: Thanks for the fix! 
I'll rebase this right after #67800 is merged.

Comment (jsnajdr) [reply packages/search/src/search.tsx]: That `Search` component PR is now in `trunk`, you can rebase at any time.

File: client/server/pages/index.js [hunk 7]
```diff
@@ -127,6 +128,8 @@ function getDefaultContext( request, response, entrypoint = 'entry-main' ) {
 	const reduxStore = createReduxStore( getCachedState( initialReducer, 'root' ) );
 	setStore( reduxStore, getCachedState );
 
+	const queryClient = await createQueryClient();
```
Comment (jsnajdr) [inline client/server/pages/index.js]: The original version in #67573 used a shared `QueryClient` for all logged out requests, as opposed to a separate client for each request. That enabled caching responses.

One way to bring that back would be to share just the cache:
```js
const sharedCache = new QueryCache();

function handleRequest() {
  const queryClient = new QueryClient( { queryCache: sharedCache } );
  /* ... */
}
```
I'm not sure what other internal state does a `QueryClient` have, in addition to the cache.

Comment (claudiucelfilip) [reply client/server/pages/index.js]: That looks like a better approach the original version.

I was hoping to restore the cache with the next PR. We should also consider only caching logged-out users.
For the next PR, I'd also like to include react-query hydration

**Comment by noahtallen** [comment]
> We should also consider only caching logged-out users.

I have a PR up here which effectively disables the SSR pipeline for logged-in requests: https://github.com/Automattic/wp-calypso/pull/67837. Looking for feedback on it, but I also wonder if it will automatically improve the caching situation for logged-in cases.

File: client/webpack.config.node.js [hunk 8]
```diff
@@ -163,6 +163,7 @@ const webpackConfig = {
 			BUILD_TIMESTAMP: JSON.stringify( new Date().toISOString() ),
 			COMMIT_SHA: JSON.stringify( commitSha ),
 			'process.env.NODE_ENV': JSON.stringify( bundleEnv ),
+			__i18n_text_domain__: JSON.stringify( 'default' ),
```
Comment (tyxla) [inline client/webpack.config.node.js:166]: Do we need the `JSON.stringify` here?

File: client/lib/plugins/sanitize-section-content.js [hunk 9]
```diff
@@ -132,6 +132,10 @@ const replacementFor = ( node ) => {
  * @returns {string} sanitized HTML
  */
 export const sanitizeSectionContent = ( content ) => {
+	// @TODO: replace with server-side sanitizer
+	if ( 'undefined' === typeof DOMParser ) {
```
Comment (tyxla) [inline client/lib/plugins/sanitize-section-content.js]: What are we planning to do here? I'd expect we handle sanitization somehow, otherwise bad bad things could happen?

File: client/data/marketplace/use-wpcom-plugins-query.ts [hunk 10]
```diff
@@ -37,31 +37,39 @@ const fetchWPCOMPlugins = ( type: Type, searchTerm?: string, tag?: string ) => {
...
-		getCacheKey( type + searchTerm + tag + `${ enabled ? 'enabled' : 'disabled' }` ),
-		() => fetchWPCOMPlugins( type, searchTerm, tag ),
-		{
-			select: ( data ) => normalizePluginsList( data.results ),
-			enabled: enabled,
-			staleTime: staleTime,
-			refetchOnMount: refetchOnMount,
-		}
-	);
+	return useQuery( ...getWPCOMPluginsQueryParams( type, searchTerm, tag ), {
```
Comment (tyxla) [inline client/data/marketplace/use-wpcom-plugins-query.ts:65]: Seems like the query and TS improvements could be part of their own PR.

File: client/my-sites/plugins/controller-logged-out.js [hunk 11]
```diff
@@ -0,0 +1,73 @@
+import {
+	getWPCOMFeaturedPluginsQueryParams,
+	getWPCOMPluginsQueryParams,
+} from 'calypso/data/marketplace/use-wpcom-plugins-query';
+import { getWPORGPluginsQueryParams } from 'calypso/data/marketplace/use-wporg-plugin-query';
+import { requestProductsList } from 'calypso/state/products-list/actions';
+import { getProductsList } from 'calypso/state/products-list/selectors';
+
+function getProps( context ) {
```
Comment (tyxla) [inline client/my-sites/plugins/controller-logged-out.js]: Too broad name?

File: client/my-sites/plugins/controller-logged-out.js [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+	const props = {
+		path: context.path,
+		locale: context.lang,
+		tag: '',
+	};
+	return props;
```
Comment (tyxla) [inline client/my-sites/plugins/controller-logged-out.js]: Nit: this could be shortened: 

```suggestion
function getProps( { path, lang } ) {
	return {
		path,
		locale: lang,
		tag: '',
	};
```

**Review by tyxla — COMMENTED** [review]
I think we could split this into multiple PRs with different purposes. I'm heading AFK next week too, so I'll defer the final review to @jsnajdr and @noahtallen

File: client/webpack.config.node.js (hunk 8 above)
Comment (claudiucelfilip) [reply client/webpack.config.node.js:166]: This was taken from the [client webpack.config](https://github.com/Automattic/wp-calypso/blob/trunk/client/webpack.config.js#L295) and is similar to the [config for editing-toolkit](https://github.com/Automattic/wp-calypso/blob/trunk/apps/editing-toolkit/webpack.config.js#L86). 
It needs to use `'"default"'` to replace `__i18n_text_domain__`.

File: client/lib/plugins/sanitize-section-content.js (hunk 9 above)
Comment (claudiucelfilip) [reply client/lib/plugins/sanitize-section-content.js]: The plan is to follow up with the sanitizer changes after this PR.
Normally, this should never happen since the SSR routes are restricted behind the config flag. I've added another flag check so it's clearer and avoids any uncertainty. 

Once we have content sanitizing and react-query cache and hydration working, we should be safe to switch the flag.

File: client/data/marketplace/use-wpcom-plugins-query.ts (hunk 10 above)
Comment (claudiucelfilip) [reply client/data/marketplace/use-wpcom-plugins-query.ts:65]: Fair enough. I'll add this improvement in a follow-up PR for all query functions.

File: client/my-sites/plugins/controller-logged-out.js (hunk 11 above)
Comment (claudiucelfilip) [reply client/my-sites/plugins/controller-logged-out.js]: Good point, I replaced it with `getQueryOptions`.

File: client/data/marketplace/use-wpcom-plugins-query.ts [hunk 13]
```diff
@@ -37,6 +37,16 @@ const fetchWPCOMPlugins = ( type: Type, searchTerm?: string, tag?: string ) => {
 	);
 };
 
+export const getWPCOMPluginsQueryParams = (
+	type: Type,
+	searchTerm?: string,
+	tag?: string
+): [ QueryKey, QueryFunction< { results: Array< { [ plugin: string ]: Plugin } > } > ] => {
```
Comment (jsnajdr) [inline client/data/marketplace/use-wpcom-plugins-query.ts:44]: The `Plugin` type is not defined. More precisely, it points to some obscure deprecated built-in DOM type for a "browser plugin", whatever that is, not to a Calypso-defined type.

Comment (jsnajdr) [inline client/data/marketplace/use-wpcom-plugins-query.ts:44]: What if the query function dereferenced the `.results` field, and its return type was `Array<Record<string, Plugin>>`. It's a bit unfortunate that the `/marketplace/products` endpoint wraps results in a `results` field at all. It should return the array directly. It's simpler and more in line with WP REST conventions.

File: client/data/marketplace/use-wpcom-plugins-query.ts [hunk 14]
```diff
@@ -37,6 +37,16 @@ const fetchWPCOMPlugins = ( type: Type, searchTerm?: string, tag?: string ) => {
...
+	const cacheKey = getCacheKey( type + searchTerm + tag );
```
Comment (jsnajdr) [inline client/data/marketplace/use-wpcom-plugins-query.ts:45]: There should be some separator between the parts of the key, so that they are not glued together as `typesearchtermtag`. That can be ambiguous.

File: client/server/pages/index.js [hunk 15]
```diff
@@ -227,8 +231,10 @@ function getDefaultContext( request, response, entrypoint = 'entry-main' ) {
 }
 
 const setupDefaultContext = ( entrypoint ) => ( req, res, next ) => {
-	req.context = getDefaultContext( req, res, entrypoint );
-	next();
+	getDefaultContext( req, res, entrypoint ).then( ( context ) => {
+		req.context = context;
+		next();
+	} );
```
Comment (jsnajdr) [inline client/server/pages/index.js]: It's much more elegant to turn this function into an async function (`(entrypoint) => async (req, res, next) => ...`, and then the only change in the body is adding `await`:
```js
req.context = await getDefaultContext( ... );
```

**Review by jsnajdr — APPROVED** [review]
Looks mergeable for something that's hidden behind a flag. Let's continue iterating on this :+1:

File: client/server/pages/index.js [hunk 16]
```diff
@@ -131,6 +132,8 @@ function getDefaultContext( request, response, entrypoint = 'entry-main' ) {
 	const reduxStore = createReduxStore( getCachedState( initialReducer, 'root' ) );
 	setStore( reduxStore, getCachedState );
 
+	const queryClient = await createQueryClient();
```
Comment (noahtallen) [inline client/server/pages/index.js]: I'm curious if this adds any performance overhead to the server -- it run on essentially every route request (SSR or not) since `getDefaultContext` runs for every section. Doesn't seem like it should make a meaningful difference, but I'm curious :)

Comment (jsnajdr) [reply client/server/pages/index.js]: This is a good point. `getDefaultContext` should be getting smaller over time rather than bigger, as code is moved to smaller, specialized handlers. Maybe this [`getEnhancedContext`](https://github.com/Automattic/wp-calypso/blob/add/plugins-landing-ssr/client/server/isomorphic-routing/index.js#L66-L77) handler, which runs only for true SSR, would be a better place for creating the `QueryClient`.

On the other hand, constructing the `QueryClient` is a very fast operation that's unlikely to cause any observable slowdown. The [constructor](https://github.com/TanStack/query/blob/900d2c51306a91372ae38d0d6846ae61a165c326/packages/query-core/src/queryClient.ts#L68-L75) merely initializes a few fields to empty-ish values and that's all. No CPU-intensive work and no large memory allocations.

Comment (claudiucelfilip) [reply client/server/pages/index.js:135]: Looking deeper, the async part is react-query's [persistQueryClient](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/query-client.ts#L26) which doesn't do anything as long as there's no [window object](https://github.com/TanStack/query/blob/v3.21.1/src/persistQueryClient-experimental/index.ts#L51).
Double-checked and this seems to take `0s` to evaluate.

Comment (claudiucelfilip) [reply client/server/pages/index.js]: Thanks for catching this.
I moved the `createQueryClient` to [getEnhancedContext](https://github.com/Automattic/wp-calypso/blob/add/plugins-landing-ssr/client/server/isomorphic-routing/index.js#L66-L77) as suggested.

**Review by noahtallen — APPROVED** [review]
Thanks for all the improvements!
